onnxruntime/core/common/threadpool.cc

```diff
@@ -21,12 +21,222 @@ limitations under the License.
 #include "core/common/eigen_common_wrapper.h"
 #include "core/platform/EigenNonBlockingThreadPool.h"
 #include "core/platform/ort_mutex.h"
+#ifdef _WIN32
+#include "processthreadsapi.h"
+#include <codecvt>
+#include <locale>
+#elif defined(__APPLE__)
+#include <cpuid.h>
+#else
+#include <sched.h>
+#endif
 
 namespace onnxruntime {
 
 namespace concurrency {
 
-// A sharded loop counter distributes loop iterations between a set of worker threads.  The iteration space of
+ThreadPoolProfiler::ThreadPoolProfiler(int num_threads, const CHAR_TYPE* threal_pool_name) : 
+    num_threads_(num_threads) {
+  child_thread_stats_.assign(num_threads, {});
+  if (threal_pool_name) {
+#ifdef _WIN32
+    using convert_type = std::codecvt_utf8<wchar_t>;
+    std::wstring_convert<convert_type, wchar_t> converter;
+    threal_pool_name_ = converter.to_bytes(threal_pool_name);
+#else
+    threal_pool_name_ = threal_pool_name;
+#endif
+  } else {
+    threal_pool_name_ = "unnamed_thread_pool";
+  }
+}
+
+ThreadPoolProfiler::~ThreadPoolProfiler() {
+  enabled_ = false;
+}
+
+void ThreadPoolProfiler::Start() {
+  enabled_ = true;
+}
+
+ThreadPoolProfiler::MainThreadStat& ThreadPoolProfiler::GetMainThreadStat() {
+  static thread_local std::unique_ptr<MainThreadStat> stat;
+  if (!stat) {
+    stat.reset(new MainThreadStat());
+  }
+  return *stat;
+}
+
+std::string ThreadPoolProfiler::Stop() {
+  ORT_ENFORCE(enabled_, "Profiler not started yet");
+  std::stringstream ss;
+  ss << "{\"main_thread\": {"
+     << "\"thread_pool_name\": \""
+     << threal_pool_name_ << "\", "
+     << GetMainThreadStat().Reset()
+     << "}, \"sub_threads\": {"
+     << DumpChildThreadStat()
+     << "}}";
+  return ss.str();
+}
+
+void ThreadPoolProfiler::LogCoreAndBlock(std::ptrdiff_t block_size) {
+  if (enabled_) {
+    GetMainThreadStat().LogCore();
+    GetMainThreadStat().LogBlockSize(block_size);
+  }
+}
+
+void ThreadPoolProfiler::LogStart() {
+  if (enabled_) {
+    GetMainThreadStat().LogStart();
+  }
+}
+
+void ThreadPoolProfiler::LogEnd(ThreadPoolEvent evt) {
+  if (enabled_) {
+    GetMainThreadStat().LogEnd(evt);
+  }
+}
+
+void ThreadPoolProfiler::LogEndAndStart(ThreadPoolEvent evt) {
+  if (enabled_) {
+    GetMainThreadStat().LogEndAndStart(evt);
+  }
+}
+
+void ThreadPoolProfiler::MainThreadStat::LogCore() {
```

Comment:
If this is not being done by the caller -- we should make this periodic so that we are not making OS calls to get the current CPU on each loop

Reply:
Then we will have to log time point again for each loop to keep track of the "period"?
This would introduce more overhead to main thread and adding up the code complexity no?